lib/semian/lru_hash.rb

```diff
@@ -0,0 +1,91 @@
+class LRUHash
+  # This LRU (Least Recently Used) hash will allow
+  # the cleaning of resources as time goes.
+  # The  goal is to remove the least recently used resources
+  # everytime we set a new resource. A default window of
+  # 5 minutes will allow empty item to stay in the hash
+  # for a maximum of 5 minutes
+  extend Forwardable
+  def_delegators :@table, :size, :count, :empty?
+  attr_reader :table, :minimum_time_in_lru
+  MINIMUM_TIME_IN_LRU = 300
+
+  class NoopMutex
+    def synchronize(*)
+      yield
+    end
+
+    def locked?
+      false
+    end
+  end
+
+  [:keys, :clear].each do |attribute|
+    define_method :"#{attribute}" do
+      @lock.synchronize { @table.public_send(attribute) }
+    end
+  end
+
+  def initialize
+    @minimum_time_in_lru = MINIMUM_TIME_IN_LRU
+    @table = {}
+    @lock =
+      if Semian.thread_safe?
+        Mutex.new
+      else
+        NoopMutex.new
+      end
+  end
+
+  def set(key, resource)
```

Comment:
I'm not convinced this moving of the mutex is correct. The delete + set are no longer atomic.

Can you explain how you're convinced this is correct?